src/yara.cc

```diff
@@ -952,19 +953,118 @@ int scanCallback(int message, void* data, void* param) {
+
+	if (! rules_compiled) {
+		Nan::ThrowError("Please call configure() before reconfigureRules()");
+		return;
+	}
+	Local<Object> options = Nan::To<Object>(info[0]).ToLocalChecked();
+
+	Local<Array> variables = Local<Array>::Cast(
+			Nan::Get(options, Nan::New("variables").ToLocalChecked()).ToLocalChecked()
+		);
+
+	for (uint32_t i = 0; i < variables->Length(); i++) {
+		if (Nan::Get(variables, i).ToLocalChecked()->IsObject()) {
+			Local<Object> variable = Nan::To<Object>(Nan::Get(variables, i).ToLocalChecked()).ToLocalChecked();
+
+			VarType type;
+			std::string id;
+
+			Local<Uint32> t = Nan::To<Uint32>(Nan::Get(variable, Nan::New("type").ToLocalChecked()).ToLocalChecked()).ToLocalChecked();
+			type = (VarType) t->Value();
+
+			Local<String> i = Nan::To<String>(Nan::Get(variable, Nan::New("id").ToLocalChecked()).ToLocalChecked()).ToLocalChecked();
+			id = *Nan::Utf8String(i);
+
+			VarConfig* var_config = new VarConfig();
+
+			var_config->type = type;
+			var_config->id = id;
+
+			int rc;
+
+			switch (type) {
+				case IntegerVarType:
+					var_config->value_integer = Nan::To<Integer>(Nan::Get(variable, Nan::New("value").ToLocalChecked()).ToLocalChecked()).ToLocalChecked()->Value();
+					rc = yr_rules_define_integer_variable(
+							scanner->rules,
+							var_config->id.c_str(),
+							var_config->value_integer
+						);
+					if (rc != ERROR_SUCCESS)
+						yara_throw(YaraError, "yr_rules_define_integer_variable() failed: "
+								<< getErrorString(rc));
+					break;
+				case FloatVarType:
+					var_config->value_float = Nan::To<Number>(Nan::Get(variable, Nan::New("value").ToLocalChecked()).ToLocalChecked()).ToLocalChecked()->Value();
+					rc = yr_rules_define_float_variable(
+							scanner->rules,
+							var_config->id.c_str(),
+							var_config->value_float
+						);
+					if (rc != ERROR_SUCCESS)
+						yara_throw(YaraError, "yr_rules_define_float_variable() failed: "
+								<< getErrorString(rc));
+					break;
+				case BooleanVarType:
+					var_config->value_boolean = Nan::To<Boolean>(Nan::Get(variable, Nan::New("value").ToLocalChecked()).ToLocalChecked()).ToLocalChecked()->Value();
+					rc = yr_rules_define_boolean_variable(
+							scanner->rules,
+							var_config->id.c_str(),
+							var_config->value_boolean
+						);
+					if (rc != ERROR_SUCCESS)
+						yara_throw(YaraError, "yr_rules_define_boolean_variable() failed: "
+								<< getErrorString(rc));
+					break;
+				case StringVarType:
+					var_config->value_string = *Nan::Utf8String(Nan::To<String>(Nan::Get(variable, Nan::New("value").ToLocalChecked()).ToLocalChecked()).ToLocalChecked());
+					rc = yr_rules_define_string_variable(
+							scanner->rules,
+							var_config->id.c_str(),
+							var_config->value_string.c_str()
+						);
+					if (rc != ERROR_SUCCESS)
+						yara_throw(YaraError, "yr_rules_define_string_variable() failed: "
+								<< getErrorString(rc));
+					break;
+			}
+		}
+	}
+
+	scanner->unlock();
```

Comment:
Should this (and the other added one) be in `finally` blocks? Does `yara_throw` actually throw in the C++ too, or does it result in an exception once the method returns?

Reply:
Good catch! I should have done that, I'll fix it.

Reply:
There are no finally blocks in C++. After doing some reading it seems people usually set those sort of calls in the destructor so it gets called after the object goes out of scope, however the scanner is long lived so we can't rely on that. I added it on the catch and after the try.

Reply:
I don't know what the `ScannerWrap` type is, but is it an option to have it return a lock instead of mutating the object? That way you can get rid of both the calls to `scanner->unlock()` and the try/catch-block. (Provided that the returned lock implements RIAA of course, but it should in any case.)

Reply:
That's a great idea! I was reading about RIAA last night (I'm learning C++ as I go 😬 ). I'll play around with it today and see if I can get it working. At first I thought about implementing RIAA in the ScannerWrap type, but because of it's life cycle that's not possible for the lock, it hadn't occurred me of using a lock type that implemented it instead! Thanks!

Reply:
Yes, a scope-based lock like that is the typical C++ way to make sure these are handled properly. If we can do it here, great.

Reply:
@miguelxpn What you're looking for is probably [std::lock_guard](https://devdocs.io/cpp/thread/lock_guard) and/or some of the other standard locks in the C++ stdlib. They do it all for you, and there's really no reason to use anything else.

Reply:
I see you already did that :) Yay! :)